External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,507 @@
+
+      ContainerType::iterator ClaimBufferImpl(size_t Size) {
+        auto BuffersEnd = OwnedBuffers.end();
+        ContainerType::iterator BestFit = BuffersEnd;
+        ContainerType::iterator UnsizedFit = BuffersEnd;
+
+        auto Now = ClockType::now();
+        {
+          // Spin the non-owned buffers and see if we can take ones past the period
+          for (auto it = NonOwnedBuffers.begin(); it != NonOwnedBuffers.end();) {
+            // 1) Can't take anything that the client has still claimed
+            // 2) Needs to still be last used beyond our time threshold
+            // 3) Only take the oldest buffer
+            if ((*it)->CurrentClientOwnedFlag->load() == ClientFlags::FLAG_DISOWNED) {
+              auto UsedTime = (*it)->LastUsed.load(std::memory_order_relaxed);
+              if ((Now - UsedTime) >= DURATION) {
+                ClientFlags Expected = ClientFlags::FLAG_DISOWNED;
+                if ((*it)->CurrentClientOwnedFlag->compare_exchange_strong(Expected, ClientFlags::FLAG_FREE)) {
+                  // We managed to take away ownership
+                  // Put it back in the regular pool and come back to it
+                  (*it)->CurrentClientOwnedFlag = nullptr;
+                  OwnedBuffers.emplace_back(*it);
+                  it = NonOwnedBuffers.erase(it);
+                  continue;
+                }
+              }
+            }
+
+            ++it;
+          }
+        }
+
+        {
+          // Walk all the allocations and find a buffer that fits
+          for (auto it = OwnedBuffers.begin(); it != BuffersEnd; ++it) {
+            if ((*it)->Size == Size) {
+              BestFit = it;
+              break;
+            }
+
+            if ((*it)->Size > Size) {
+              UnsizedFit = it;
+            }
+          }
+
+          // If we didn't have an exact fit then use an unsized fit
+          if (BestFit == BuffersEnd) {
+            BestFit = UnsizedFit;
+          }
+
+          {
+            std::chrono::time_point<ClockType> LRUTime{};
+            ContainerType::iterator LastUsed = BuffersEnd;
+
+            // Walk all the allocations and find a buffer to erase
+            for (auto it = OwnedBuffers.begin(); it != OwnedBuffers.end(); ++it) {
+              // Ensure that the LRU value is past our duration threshold and isn't the one we are claiming
+              // Also only select a single memory region
+              if (it != BestFit) {
+                auto UsedTime = (*it)->LastUsed.load(std::memory_order_relaxed);
+                if ((Now - UsedTime) >= DURATION &&
+                    UsedTime > LRUTime) {
+                  LastUsed = it;
+                  LRUTime = UsedTime;
+                }
+              }
+            }
+
+            // If we found a buffer then free it
+            if (LastUsed != BuffersEnd) {
+              Free((*LastUsed)->Ptr, (*LastUsed)->Size);
+              delete *LastUsed;
+              OwnedBuffers.erase(LastUsed);
```

Comment:
`it = OwnedBuffers.erase(LastUsed);` ?

Reply:
This is no longer isn't the for loop, there is no iterator to update.

Reply:
Misread the indentation. So at most one buffer gets free'd for every call to `ClaimBufferImpl`. Is this to avoid stutter?

Reply:
Yes, only one unclaimed buffer gets freed and removed from the `UnclaimedBuffers` container each time a new claim comes in. This is setup this way for longer running applications having mostly idle threads and then a handful of active threads will slowly chip away at the problem.
Also if you have an intermediate mix of active and inactive, you aren't suddenly getting a spike of deleted unclaimed buffers and new threads having absolutely no buffers available.

Basically the five second buffer of only freeing one buffer at a time is a nice sponge in both directions.